Refactor `copy` API in async/await

SukkaW opened the pull request:
The PR is the first part of proposal #1020:

- The PR only migrates the `copy` API. Other APIs will be migrated in the following PRs
- The PR adds Promise-based implementations for some utility functions. The callback-based implementations are retained, but they will be removed once the rest of the library has migrated to Promise-based implementations.

SukkaW commented:
@RyanZim The PR is ready to be reviewed now. `npm run lint` passes, and all test cases pass as well, check CI run here: https://github.com/SukkaW/node-fs-extra/actions/runs/6582727479

SukkaW commented on lib/copy/__tests__/copy-broken-symlink.test.js:
```diff
@@ -5,7 +5,7 @@ const os = require('os')
 const path = require('path')
 const assert = require('assert')
-const copy = require('../copy')
+const { copy } = require('../')
```
I need to change the import here. `lib/copy/copy.js` now exports the promise-only API (instead of the callback-only API), while `lib/copy/index.js` now exports the unified API.

SukkaW commented on lib/util/__tests__/utimes.test.js:
```diff
@@ -33,7 +33,7 @@ describe('utimes', () => {
     // reset stubs
     gracefulFsStub = {}
-    utimes = proxyquire('../utimes', { 'graceful-fs': gracefulFsStub })
+    utimes = proxyquire('../utimes', { '../fs': gracefulFsStub })
```
The mock here has to be replaced, as `lib/util/utimes.js` no longer uses `graceful-fs` directly. It now uses the unified API from `lib/fs/index.js`.

RyanZim commented on lib/copy/copy.js:
```diff
@@ -30,209 +26,188 @@ function copy (src, dest, opts, cb) {
-      return setDestTimestampsAndMode(srcMode, src, dest, cb)
-    })
+    await makeFileWritable(dest, srcMode)
+    return setDestTimestampsAndMode(srcMode, src, dest)
```
No need to duplicate this inside the conditional since we're not inside a callback anymore; just allow it to fall through to the `setDestTimestampsAndMode` call below.

SukkaW replied:
Fixed in [`bc8ee0b` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/bc8ee0bf340f04fad0fed408bbdd0cc3747e9d0c).

RyanZim commented on lib/copy/copy.js:
```diff
@@ -30,209 +26,188 @@ function copy (src, dest, opts, cb) {
-    if (err) return cb(err)
-    return utimesMillis(dest, updatedSrcStat.atime, updatedSrcStat.mtime, cb)
-  })
+  // (See https://nodejs.org/api/fs.html#fs$stat_time_values)
```
URL change here is accidental? New link is broken for me; old one still works.

SukkaW replied:
Fixed in [`bc8ee0b` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/bc8ee0bf340f04fad0fed408bbdd0cc3747e9d0c).

RyanZim commented on lib/copy/copy.js:
```diff
@@ -30,209 +26,188 @@ function copy (src, dest, opts, cb) {
-  if (!destStat) return mkDirAndCopy(srcStat.mode, src, dest, opts, cb)
-  return copyDir(src, dest, opts, cb)
+function onDir (srcStat, destStat, src, dest, opts) {
+  if (!destStat) return mkDirAndCopy(srcStat.mode, src, dest, opts)
+  return copyDir(src, dest, opts)
```
Can reduce code duplication by inlining `mkDirAndCopy` like this:
```suggestion
  if (!destStat) await fs.mkdir(dest)
  await copyDir(src, dest, opts)
  if (!destStat) await setDestMode(dest, srcMode)
```

SukkaW replied:
Fixed in [`bc8ee0b` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/bc8ee0bf340f04fad0fed408bbdd0cc3747e9d0c).

RyanZim commented on lib/copy/copy.js:
```diff
@@ -30,209 +26,188 @@ function copy (src, dest, opts, cb) {
+  const items = await fs.readdir(src)
+  return copyDirItems(items, src, dest, opts)
 }
 
-function copyDirItems (items, src, dest, opts, cb) {
+function copyDirItems (items, src, dest, opts) {
   const item = items.pop()
-  if (!item) return cb()
-  return copyDirItem(items, item, src, dest, opts, cb)
+  if (!item) return
+  return copyDirItem(items, item, src, dest, opts)
 }
 
-function copyDirItem (items, item, src, dest, opts, cb) {
+async function copyDirItem (items, item, src, dest, opts) {
```
This logic needs to be rethought/reworked. We currently use a recursive `copyDirItems` call, which mutates `items`. This made sense to avoid the complexity of serially iterating a callback-based process, but is non-idiomatic in an async/await world. I'm thinking we use a `for` loop in `copyDir` to iterate over the items...I'll leave the exact details to you.

SukkaW replied:
Refactored this part in [`cf3aed7` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/cf3aed7a1d413e329fa7edede323f2479f3901a0).

The `onDir` now uses a for loop to iterate every item inside the directory (from `await fs.readdir`), perform the pre-copy check (the `filter` option and the `stat.checkPathsAsync`), and then invoke the `getStatsAndPerformCopy` function (which would recursively call `onDir` again).

RyanZim commented on lib/copy/copy.js:
```diff
@@ -30,209 +26,188 @@ function copy (src, dest, opts, cb) {
+  try {
+    resolvedDest = await fs.readlink(dest)
+  } catch (e) {
+    if (e.code === 'EINVAL' || e.code === 'UNKNOWN') return fs.symlink(resolvedSrc, dest)
```
Please preserve explanatory comments from original code

SukkaW replied:
Fixed in [`bc8ee0b` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/bc8ee0bf340f04fad0fed408bbdd0cc3747e9d0c).

RyanZim commented on lib/copy/copy.js:
```diff
@@ -30,209 +26,188 @@ function copy (src, dest, opts, cb) {
+    throw new Error(`Cannot overwrite '${resolvedDest}' with '${resolvedSrc}'.`)
+  }
+
+  return copyLink(resolvedSrc, dest)
+}
+
+async function copyLink (resolvedSrc, dest) {
+  await fs.unlink(dest)
+  return fs.symlink(resolvedSrc, dest)
 }
```
Can inline this logic:
```suggestion
  await fs.unlink(dest)
  return fs.symlink(resolvedSrc, dest)
}
```

SukkaW replied:
Fixed in [`bc8ee0b` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/bc8ee0bf340f04fad0fed408bbdd0cc3747e9d0c).

And the entire `copyLink` has been inlined into `onLink` as well.

RyanZim reviewed (changes requested):
I've left a several code comments. Additionally, I have a few other concerns:

1. `copy` is not expected to return a value, yet we use `return` throughout the code for final async calls inside functions; this could result in unintentionally returning a value in the future; perhaps we should use `await` instead?
2. Is there a reason for duplicating the util functions vs. just wrapping them in universalify? We could wrap them in `fromCallback` here, then port them to promise implementations and wrap in `fromPromise` in a separate PR. Anything I'm missing? (I have not reviewed those changes yet)

SukkaW commented on lib/copy/copy.js:
```diff
@@ -1,21 +1,17 @@
   }
 
-  cb = cb || function () {}
   opts = opts || {}
```
Done in [`bc8ee0b` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/bc8ee0bf340f04fad0fed408bbdd0cc3747e9d0c).

I changed to use the default assign syntax (no nested condition!):

```js
// If `opts` is not specified or undefined, assign {}
async function copy (src, dest, opts = {}) {
  // now re-assign `opts` if it is a function
  if (typeof opts === 'function') {
    opts = { filter: opts }
  }
```

SukkaW commented on lib/copy/copy.js:
```diff
@@ -30,209 +26,188 @@ function copy (src, dest, opts, cb) {
+  if (dirExists) return getStats(destStat, src, dest, opts)
+
+  const parentDirExists = await pathExists(destParent)
+  if (parentDirExists) return getStats(destStat, src, dest, opts)
```
Fixed in [`bc8ee0b` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/bc8ee0bf340f04fad0fed408bbdd0cc3747e9d0c).

SukkaW commented on lib/copy/copy.js:
```diff
@@ -30,209 +26,188 @@ function copy (src, dest, opts, cb) {
-  return mayCopyFile(srcStat, src, dest, opts, cb)
+function onFile (srcStat, destStat, src, dest, opts) {
+  if (!destStat) return copyFile(srcStat, src, dest, opts)
+  return mayCopyFile(srcStat, src, dest, opts)
```
Fixed in [`bc8ee0b` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/bc8ee0bf340f04fad0fed408bbdd0cc3747e9d0c). The `mayCopyFile` is now inlined.

SukkaW commented on lib/util/__tests__/utimes.test.js:
```diff
@@ -68,25 +70,25 @@ describe('utimes', () => {
       const fakeFd = Math.random()
 
-      gracefulFsStub.open = (pathIgnored, flagsIgnored, modeIgnored, callback) => {
+      gracefulFsStub.open = u((pathIgnored, flagsIgnored, modeIgnored, callback) => {
```
The `gracefulFsStub` needs to be universalify as the `utimesMillis` function is now universalised.

SukkaW commented:
@RyanZim I've addressed all the code review suggestions. I've also universalized the related utility functions to reduce the duplicated implementations. The CI still passes after the latest changes: https://github.com/SukkaW/node-fs-extra/actions/runs/6598480115

Note that the `coprDir` has been heavily refactored (no more `items` array mutation!).

RyanZim commented on lib/copy/copy.js:
```diff
@@ -122,43 +108,47 @@ async function setDestTimestampsAndMode (srcMode, src, dest) {
-  return copyDirItem(items, item, src, dest, opts)
-}
+  // loop through the files in the current directory to copy everything
+  for (let i = 0, len = items.length; i < len; i++) {
+    const item = items[i]
```
Why not use a `for..of` loop here? Sorry if my initial review suggested I wanted a plain `for` loop.

SukkaW replied:
Changed in [`1ce5d34` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/1ce5d348c426ea20c5006389477e163211898b22).

RyanZim commented on lib/util/utimes.js:
```diff
@@ -3,7 +3,7 @@
 const u = require('universalify').fromPromise
 
-async function utimesMillis (path, atime, mtime) {
+async function utimesMillis (path, atime, mtime, cb) {
```
Why `cb` here?

SukkaW replied:
Removed in [`1ce5d34` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/1ce5d348c426ea20c5006389477e163211898b22).

RyanZim commented on lib/copy/copy.js:
```diff
@@ -30,209 +24,156 @@ function copy (src, dest, opts, cb) {
-  })
-}
+async function onDir (srcStat, destStat, src, dest, opts) {
+  // the dest direcotry might not exist, create it
```
```suggestion
  // the dest directory might not exist, create it
```

SukkaW replied:
Typo fixed in [`1ce5d34` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/1ce5d348c426ea20c5006389477e163211898b22).

RyanZim commented on lib/util/utimes.js:
```diff
@@ -1,17 +1,33 @@
-  })
+  const fd = await fs.open(path, 'r+')
+
+  let futimesErr = null
+  try {
+    await fs.futimes(fd, atime, mtime)
+  } catch (e) {
+    futimesErr = e
+  }
+
+  let closeErr = null
+
+  try {
+    await fs.close(fd)
+  } catch (e) {
+    closeErr = e
+  }
+
+  if (futimesErr) {
+    throw futimesErr
+  }
```
We can achieve the same behavior a little more simply with `try..finally`.
```suggestion
  let closeErr
  try {
    await fs.futimes(fd, atime, mtime)
  } finally {
    // put this in a try..catch so the error from futimes is thrown preferentially to any error here
    try {
      await fs.close(fd)
    } catch (e) {
      closeErr = e
    }
  }
```

SukkaW replied:
Changed in [`1ce5d34` (#1021)](https://github.com/jprichardson/node-fs-extra/pull/1021/commits/1ce5d348c426ea20c5006389477e163211898b22).

RyanZim reviewed (changes requested):
Just a couple comments
